lib/galaxy/model/util.py

```diff
@@ -3,31 +3,94 @@
 """
 
 
+def count_toward_deleted_disk_usage(dataset_id, history_deleted, hda_deleted, dataset_deleted, dataset_disk_usage, **ids_dict):
+    """
+    ids_dict: stores the state of dataset and its associated hda and history, key=str(dataset_id), value could be one of following.
+        -1: dataset deleted
+        0: dataset not deleted, history deleted, hda deleted
+        1: dataset not deleted, history deleted, hda not deleted
+        2: dataset not deleted, history not deleted, hda deleted
+        3: dataset not deleted, history not deleted, hda not deleted
+
+    Return:
+    disk usage counted toward deleted disk usage for the input dataset
+    """
+    id_key = str(dataset_id)
+    # already counted, state fixed
+    if id_key in ids_dict and (ids_dict[id_key] == -1 or ids_dict[id_key] == 3):
+        return 0
+    if id_key not in ids_dict:       # first count; count all possibles
+        if dataset_deleted:
+            ids_dict[id_key] = -1    # must count
+            return dataset_disk_usage
+        elif hda_deleted:
+            if history_deleted:      # possible
+                ids_dict[id_key] = 0
+                return dataset_disk_usage
+            else:                    # possible
+                ids_dict[id_key] = 2
+                return dataset_disk_usage
+        else:
+            if history_deleted:     # possible
+                ids_dict[id_key] = 1
+                return dataset_disk_usage
+            else:                   # impossible
```

Comment:
If this case is impossible, should we log something to detect if it has ever happened?

Reply:
line 22, `if id_key not in ids_dict:       # first count; count all possibles`, guarantees it's the first occurrence.

Reply:
Not sure I follow, maybe I misunderstood the meaning of the # impossible comment.  If this condition is impossible to visit (or, rather, should be impossible to visit), then can it either log an error if it ever is, or be removed?

Reply:
Not 'impossible to visit'. I just mean that the dataset should NOT be counted toward deleted disk usage.  It's at the first occurrence. We need add it to ids_dict and avoid counting for later occurrences. Maybe I need change the log to something else. Any suggestion?

Reply:
Got it!  I misunderstood the comment.  Can you just change `#impossible` to something more descriptive, then, and that should handle it?

Reply:
Good suggestion!